src/main/java/org/springframework/data/redis/cache/RedisCache.java

```diff
@@ -720,20 +720,27 @@ public RedisCachePutIfAbsentCallback(BinaryRedisCacheElement element, RedisCache
 		public byte[] doInRedis(BinaryRedisCacheElement element, RedisConnection connection) throws DataAccessException {
 
 			waitForLock(connection);
-			byte[] resultValue = put(element, connection);
 
-			if (nullSafeEquals(element.get(), resultValue)) {
+			byte existingValue[] = null;
+
+			boolean keyMaintenance;
+
+			byte[] keyBytes = element.getKeyBytes();
+			byte[] value = element.get();
+
+			if (connection.setNX(keyBytes, value)) {
+				keyMaintenance = true;
+			} else {
+				existingValue = connection.get(keyBytes);
+				keyMaintenance = ObjectUtils.nullSafeEquals(value, existingValue);
```

Comment:
Not sure about the reason for expiry/key maintenance if the key was not set but the same content.